spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -19,263 +19,64 @@
-    return this;
-  }
-
-  /**
-   * Pass a row Expression to filter DataFiles to be rewritten. Note that all files that may contain data matching the
-   * filter may be rewritten.
-   *
-   * @param expr Expression to filter out DataFiles
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction filter(Expression expr) {
-    this.filter = Expressions.and(filter, expr);
-    return this;
-  }
-
   @Override
   public RewriteDataFilesActionResult execute() {
-    CloseableIterable<FileScanTask> fileScanTasks = null;
-    try {
-      fileScanTasks = table.newScan()
-          .caseSensitive(caseSensitive)
-          .ignoreResiduals()
-          .filter(filter)
-          .planFiles();
-    } finally {
-      try {
-        if (fileScanTasks != null) {
-          fileScanTasks.close();
-        }
-      } catch (IOException ioe) {
-        LOG.warn("Failed to close task iterable", ioe);
-      }
-    }
-
-    Map<StructLikeWrapper, Collection<FileScanTask>> groupedTasks = groupTasksByPartition(fileScanTasks.iterator());
-    Map<StructLikeWrapper, Collection<FileScanTask>> filteredGroupedTasks = groupedTasks.entrySet().stream()
-        .filter(kv -> kv.getValue().size() > 1)
-        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    Map<StructLikeWrapper, Collection<FileScanTask>> filteredGroupedTasks = getFilteredGroupedTasks();
 
     // Nothing to rewrite if there's only one DataFile in each partition.
     if (filteredGroupedTasks.isEmpty()) {
       return RewriteDataFilesActionResult.empty();
     }
 
     // Split and combine tasks under each partition
-    List<CombinedScanTask> combinedScanTasks = filteredGroupedTasks.values().stream()
-        .map(scanTasks -> {
-          CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
-              CloseableIterable.withNoopClose(scanTasks), targetSizeInBytes);
-          return TableScanUtil.planTasks(splitTasks, targetSizeInBytes, splitLookback, splitOpenFileCost);
-        })
-        .flatMap(Streams::stream)
-        .collect(Collectors.toList());
+    List<CombinedScanTask> combinedScanTasks = getCombinedScanTasks(filteredGroupedTasks);
 
     JavaRDD<CombinedScanTask> taskRDD = sparkContext.parallelize(combinedScanTasks, combinedScanTasks.size());
 
-    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
-    Broadcast<EncryptionManager> encryption = sparkContext.broadcast(encryptionManager);
+    Broadcast<FileIO> io = sparkContext.broadcast(SparkUtil.serializableFileIO(this.table()));
+    Broadcast<EncryptionManager> encryption = sparkContext.broadcast(this.table().encryption());
 
-    RowDataRewriter rowDataRewriter = new RowDataRewriter(table, spec, caseSensitive, io, encryption);
+    RowDataRewriter rowDataRewriter =
+        new RowDataRewriter(this.table(), this.table().spec(), isCaseSensitive(), io, encryption);
 
     List<DataFile> addedDataFiles = rowDataRewriter.rewriteDataForTasks(taskRDD);
```

Comment:
For my understanding,  this is only the different part between flink, spark, or other engines.  How about introducing  an abstracted method named `List<DataFile> rewriteDataForTasks(List<CombinedScanTask> combinedScanTask, .. )`  in this base rewrite action class. 

Then both spark and flink only need to implement this `rewriteDataForTasks` method,  that will make the code really simpler.

Reply:
thanks for your suggestion,I extract execute method to RewriteDataFilesActionBase, and add a abstract method rewriteDataForTasks in RewriteDataFilesActionBase.